dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AbstractHostedEntryResource.java

```diff
@@ -425,16 +432,19 @@ void persistSourceFiles(U tag, Collection<SourceFile> sourceFiles) {
     /**
      * Calculates the next version name. Currently assumes versions are always stringified numbers, and returns
      * the highest number + 1 as a string. Need to update when we support arbitrary names for the version.
-     * @param versions
-     * @return
+     * @param versions  The existing list of versions
+     * @return          The version name of the next version
      */
-    String calculateNextVersionName(Set<U> versions) {
-        if (versions.isEmpty()) {
+    String calculateNextVersionName(Set<U> versions, T entry) {
+        Set<String> blacklistedVersionNames = entry.getBlacklistedVersionNames();
+        Set<String> currentVersionNames = versions.stream().map(Version::getName).collect(Collectors.toSet());
+        Set<String> combinedVersionNames = Stream.concat(blacklistedVersionNames.stream(), currentVersionNames.stream()).collect(Collectors.toSet());
+        if (combinedVersionNames.isEmpty()) {
             return "1";
         } else {
-            U versionWithTheLargestName = versionWithLargestName(versions);
-            return (String.valueOf(Integer.parseInt(versionWithTheLargestName.getName()) + 1));
-
+            Set<Integer> versionNamesAsIntegers = combinedVersionNames.stream().map(Integer::parseInt).collect(Collectors.toSet());
```

Comment:
It was like this before, but looking at it now I see a possibility of two identical version names being created. An extremely unlikely event, but since you're in the code, I want to make sure I understand it right before creating an issue -- if two requests come in at once to create a version they both might try to create a new version with the same number.

I think we should have a unique index on the name column in workflowversion to prevent it. Or am I missing something?

Reply:
I think the tricky bit is that a unique index on workflowversion wouldn't quite work since you can have the same `v1` on difference workflows, so you would need a constraint across tables (which doesn't quite exist in postgres). 

A possible workaround, migrate the name to `workflow_workflowversion` and constrain there because then you could do a unique index across workflow, version name.

Reply:
I see, this vaguely rings a bell; I think we may have discussed this before in some other context.